experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -435,6 +435,191 @@ static Value buildStateCreation(ConstantStreamOp op, RewriterBase &rewriter,
+/// %1 = call @iterators.upstream.open.0(%0) :
+///          (!nested_state) -> !nested_state
+/// %2 = llvm.insertvalue %1, %arg0[0 : index] :
+///          !llvm.struct<"iterators.filter_state", (!nested_state)>
+static Value buildOpenBody(FilterOp op, RewriterBase &rewriter,
+                           Value initialState,
+                           ArrayRef<IteratorInfo> upstreamInfos) {
+  Location loc = op.getLoc();
+  Type upstreamStateType = upstreamInfos[0].stateType;
+
+  // Extract upstream state.
+  Value initialUpstreamState =
+      createExtractValueOp(rewriter, loc, upstreamStateType, initialState, {0});
+
+  // Call Open on upstream.
+  SymbolRefAttr openFunc = upstreamInfos[0].openFunc;
+  auto openCallOp = rewriter.create<func::CallOp>(
+      loc, openFunc, upstreamStateType, initialUpstreamState);
+
+  // Update upstream state.
+  Value updatedUpstreamState = openCallOp->getResult(0);
+  Value updatedState = createInsertValueOp(rewriter, loc, initialState,
+                                           updatedUpstreamState, {0});
+
+  return updatedState;
+}
+
+/// Builds IR that consumes all elements of the upstream iterator and returns
+/// a stream of those that pass the given precicate. Pseudo-code:
+///
+/// while (nextTuple = upstream->Next()):
+///     if precicate(nextTuple):
+///         return nextTuple
+/// return {}
+///
+/// Possible output:
+///
+/// %0 = llvm.extractvalue %arg0[0 : index] :
+///          !llvm.struct<"iterators.filter_state", (!nested_state)>
+/// %1:3 = scf.while (%arg1 = %0) :
+///            (!nested_state) -> (!nested_state, i1, !element_type) {
+///   %3:3 = func.call @iterators.upstream.next.0(%arg1) :
+///              (!nested_state) -> (!nested_state, i1, !element_type)
+///   %4 = scf.if %3#1 -> (i1) {
+///     %7 = func.call @predicate(%3#2) : (!element_type) -> i1
+///     scf.yield %7 : i1
+///   } else {
+///     scf.yield %3#1 : i1
+///   }
+///   %true = arith.constant true
+///   %5 = arith.xori %4, %true : i1
+///   %6 = arith.andi %3#1, %5 : i1
+///   scf.condition(%6) %3#0, %3#1, %3#2 : !nested_state, i1, !element_type
+/// } do {
+/// ^bb0(%arg1: !nested_state, %arg2: i1, %arg3: !element_type):
+///   scf.yield %arg1 : !nested_state
+/// }
+/// %2 = llvm.insertvalue %1#0, %arg0[0 : index] :
+///          !llvm.struct<"iterators.filter_state", (!nested_state)>
+static llvm::SmallVector<Value, 4>
+buildNextBody(FilterOp op, RewriterBase &rewriter, Value initialState,
+              ArrayRef<IteratorInfo> upstreamInfos, Type elementType) {
+  Location loc = op.getLoc();
+
+  // Extract upstream state.
+  Type upstreamStateType = upstreamInfos[0].stateType;
+  Value initialUpstreamState =
+      createExtractValueOp(rewriter, loc, upstreamStateType, initialState, {0});
+
+  // Main while loop.
+  Type i1 = rewriter.getI1Type();
+  SmallVector<Type> nextResultTypes = {upstreamStateType, i1, elementType};
+  scf::WhileOp whileOp = scf::createWhileOp(
```

Comment:
As a followup you could start using ImplicitLocOpBuilder and call it just `b`.
Some if these would become 1-liners and generally improve legibility I think.

Reply:
Aha, I see, yes, that would avoid having to carry `loc` around everywhere and shorten a few lines.

Before I can do that, I'd have to change the `build{Open,Next,Close}Body` function signatures, because they currently carry a `RewriterBase`, which `ImplicitLocOpBuilder` doesn't derive from. But I think that that's a good idea anyways: I *think* I don't need the rewriter functionally (and thus the functions are called `build*` and not `rewrite*`).

How would you handle the `scf::createWhileOp` helper and similar that we created? They (1) take a `Location` argument and (2) take function arguments with a `Location` in their signature. Add a variant without `Location` argument in both places?